### PR TITLE
Saving User scripts updates subscribed panels

### LINKED ISSUE #4989: User script changes do not update until another message arrives
**Description**
I've created a layout with two raw message panels and a user script panel. I've written a basic user script using the default "new script" template which outputs `hello: world`. My script has one input topic and one output topic. One raw message panel is subscribed to my input topic and one to my output topic.

I now load a data source that contains messages on my input topic. I play through the data some and pause. The raw message panel shows me the expected output `hello: world`. I update the user script to change the string `world` to something else and click `save`. My output raw message panel does not update with the new message. If I start to play the output raw message panel finally updates.

**Expected Behavior**
The output raw message panel updates with the new message. I believe this behavior use to work (it does seem to work with webviz).

Maybe we've changed something in our user script panel or in our message pipeline that does not allow us to re-transmit the same "frame" again but it is worth investigating why we regressed.

### PR DESCRIPTION
**Pull request by snosenzo**
**User-Facing Changes**

- Saving a user script will update subscribed panels with new output messages

**Description**

 Added state to remember last of each topic's message in the userNodePlayer and every time `setUserNodes` is called as an update to an existing node, that node will be added to a set of nodes that need to have their outputs refreshed. In the next `_onPlayerState` call this will read in the previous messages for each input topic and add them to the messages to be processed in `_getMessages` and then remove them from the list of nodes to update. `IterablePlayer` triggers this during `idle` state. This avoids an additional `_emitState` call and issues from my previous solutions with race conditions where previous playerStates were being overridden and old sourceCode was being shown.

Fixes #4989
FG-1019

### CODE REVIEW

**Review by defunctzombie — CHANGES_REQUESTED** [review]
Please let's avoid adding a requestBackfill method to the player API. I felt pretty strongly that this was a weird anti-pattern for the player interface and worked to purge it from our code in favor of the player updating when it gets subscriptions.

If the user script needs to re-process the current frame I think it should do that within its own logic.

**Comment by snosenzo** [comment]
> Please let's avoid adding a requestBackfill method to the player API. I felt pretty strongly that this was a weird anti-pattern for the player interface and worked to purge it from our code in favor of the player updating when it gets subscriptions.
> 
> If the user script needs to re-process the current frame I think it should do that within its own logic.

Ah okay, good to know. I refactored to make the `UserNodePlayer` keep track of last previous input messages and to reprocess them when the UserNodes are updated. Let me know if this isn't desirable, the last option I have in mind is to unsubscribe and re-subscribe to the input topic, but that felt a little hacky

File: packages/studio-base/src/players/UserNodePlayer/index.ts [hunk 1]
```diff
@@ -796,8 +812,37 @@ export default class UserNodePlayer implements Player {
...
+          }
+        }
+        this._userNodeIdsNeedUpdate.clear();
+
+        for (const message of messages) {
+          this._lastMessageByInputTopic.set(message.topic, message);
+        }
+
+        const messagesToBeParsed =
+          messagesForRecompute.length > 0 ? messages.concat(messagesForRecompute) : messages;
```
Comment (defunctzombie) [inline packages/studio-base/src/players/UserNodePlayer/index.ts:843]: Does this not end up with some duplicate messages? Is the assumption here that `messages` is empty? If any items in `messages` match any items in `_lastMessageByInputTopic` then we end up with duplicate messages into the _getMessages call?

**Review by defunctzombie — COMMENTED** [review]
Does this also run for preload messages nodes? So if I plot the output of a user script and then update that script, does my plot update?

**Comment by snosenzo** [comment]
@defunctzombie 
> does my plot update

confirmed that it does work with plot view

> duplicate messages?

good catch -- I added some logic to make sure input topics to get messages from were unique and that it wouldn't use old messages if a message with a topic to be updated was in active state.

**Review by defunctzombie — APPROVED** [review]
Check that this also works on preloaded messages and I think this is good to go.

**Pull request merged**